presto-main/src/main/java/com/facebook/presto/operator/WorkProcessor.java

```diff
@@ -147,25 +142,56 @@
          * finish processing and flush any remaining data.
          * @return the current transformation state, optionally bearing a result
          * <ul>
-         * <li>if state is {@link ProcessorState#finished()} then transformation has finished
+         * <li>if state is {@link #finished()} then transformation has finished
          * and the process method won't be called again.
-         * <li>if state is {@link ProcessorState#needsMoreData()} then transformation requires
+         * <li>if state is {@link #needsMoreData()} then transformation requires
          * more data in order to continue. The process method will be called with
          * a new input element or with {@link Optional#empty()} if there are no more elements.
-         * <li>if state is {@link ProcessorState#yield()} then transformation has yielded.
+         * <li>if state is {@link #yield()} then transformation has yielded.
          * The process method will be called again with the same input element.
-         * <li>if state is {@link ProcessorState#blocked(ListenableFuture)} then transformation is blocked.
+         * <li>if state is {@link #blocked(ListenableFuture)} then transformation is blocked.
          * The process method will be called again with the same input element after blocked
          * future is done
-         * <li>if state is {@link ProcessorState#ofResult(Object, boolean)} then the transformation
-         * has produced a result. If <code>needsMoreData</code> {@link ProcessorState#ofResult(Object, boolean)}
+         * <li>if state is {@link #ofResult(Object, boolean)} then the transformation
+         * has produced a result. If <code>needsMoreData</code> {@link #ofResult(Object, boolean)}
          * parameter is <code>true</code> then the process method will be called again with a new element
          * (or with {@link Optional#empty()} if there are no more elements).
          * If <code>needsMoreData</code> parameter is <code>false</code> then the process method
          * will be called again with the same input element.
          * </ul>
          */
-        ProcessorState<R> process(Optional<T> elementOptional);
+        TransformationState<R> process(Optional<T> elementOptional);
+
+        default TransformationState<T> needsMoreData()
+        {
+            return TransformationState.needsMoreData();
+        }
+
+        default TransformationState<T> blocked(ListenableFuture<?> blocked)
+        {
+            return TransformationState.blocked(blocked);
+        }
+
+        default TransformationState<T> yield()
+        {
+            return TransformationState.yield();
+        }
+
+        @Deprecated
```

Comment:
Remove. It is useful in transformations that can yield/block but map elements 1 to 1

Reply:
good catch. I wanted to remove this method, because one has to wonder whether this `needsData` or not.
Or even better, to remove both overloads, and add some meaningfully named  methods instead. Maybe `resultAndNeedsMoreData(T)`, `resultAndGiveMeTheSame(T)` (kidding, please suggest something)

Reply:
Hm. I would change the other one:
```
ofResult(T result, boolean needsMoreData)
```
to maybe:
```
ofResultAndKeepInput()
```
?

Reply:
i think:
- `result`  (current `ofResult`)
- and `partialResult` (i.e. keep input)

Reply:
Actually there is code like https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/util/MergeSortedPages.java#L132
Where ofResult needsMoreData param is used (and useful). Those are cases where we are flushing buffered data and depending if input finished we can/cannot request more data.

Maybe let's keep those method names (I don't have better proposals)

Reply:
Another similar one (flushing buffered data): https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/MergingHashAggregationBuilder.java#L121